networks/inference_net.py

```diff
@@ -0,0 +1,155 @@
+        self.conv3 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=3, padding=padding)
+        self.conv4 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=3, padding=padding)
+        self.conv5 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=3, padding=padding)
+        self.conv6 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=3, padding=padding)
+        self.out_mean = nn.Conv2d(in_channels=64, out_channels=48, kernel_size=3, padding=padding)
+        self.out_logvar = nn.Conv2d(
+            in_channels=64,
+            out_channels=out_channels,
+            kernel_size=3,
+            padding=padding
+        )
+        self.type(dtype)
+
+    def forward(self, x):
+        """Compute the encoding of the given sequence of images.
+
+        Args:
+            x (torch.Tensor): A N x S x H x W tensor containing the sequence of frames. N is the
+                batch size, S the number of frames in a sequence, H and W the height and width.
+
+        Returns:
+            A tuple z, mu, stddev, that are all N x 48 x H x W tensors. z is the latent encoding
+            for the given input sequence, while mu and stddev are the parameters of the
+            variational distribution.
+        """
+        x = self.input_conv(x)
+        x = self.conv1(x)
+        x = self.conv2(x)
+        x = self.conv3(x)
+        x = self.conv4(x)
+        x = self.conv5(x)
+        x = self.conv6(x)
+        mean = self.out_mean(x)
+        stddev = torch.exp(0.5 * self.out_logvar(x))
+        epsilon = torch.randn_like(mean)
+        z = mean + stddev * epsilon
+        return z, mean, stddev
+
+
+class TransformerNet(nn.Module):
+    """Implementation of the encoder-transformer network, that maps the latent space into the
+    phase space.
+    """
+
+    def __init__(self, in_channels, out_channels, dtype=torch.float):
+        """Instantiate the 4 convolutional layers.
+
+        Args:
+            in_channels (int): Number of input in_channels.
+            out_channels (int): Number of in_channels of the output.
+            dtype (torch.dtype): Type of the weights.
+        """
+        super().__init__()
+        self.conv1 = nn.Conv2d(
+            in_channels=in_channels,
+            out_channels=64,
+            kernel_size=3,
+            padding=1,
+            stride=2
+        )
+        self.conv2 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=3, padding=1, stride=2)
+        self.conv3 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=3, padding=1, stride=2)
+        self.out_conv = nn.Conv2d(
+            in_channels=64,
+            out_channels=out_channels,
+            kernel_size=3,
+            padding=1
+        )
+        self.type(dtype)
+
+    def forward(self, x):
+        x = self.conv1(x)
+        x = self.conv2(x)
+        x = self.conv3(x)
+        x = self.out_conv(x)
+        q, p = to_phase_space(x)
+        return q, p
+
+
+def to_phase_space(encoding):
```

Comment:
Should this be a private method of the transformer class? Or is it called other places?

Reply:
Fixed in 4b0edd6